web3/utils/datastructures.py

```diff
@@ -87,3 +89,96 @@ def __eq__(self, other):
             return self.__dict__ == dict(other)
         else:
             return False
+
+
+class NamedElementStack(Mapping):
+    def __init__(self, init_elements, valid_element=callable):
+        self._queue = OrderedDict()
+        self._callbacks = []
+        for element in reversed(init_elements):
+            if valid_element(element):
+                self.add(element)
+            else:
+                self.add(*element)
+
+    def add(self, element, name=None):
+        if name is None:
+            name = element
+
+        if name in self._queue:
+            if name is element:
+                raise ValueError("You can't add the same un-named instance twice")
+            else:
+                raise ValueError("You can't add the same name again, use replace instead")
+
+        self._queue[name] = element
+        self._changed()
+
+    def clear(self):
+        self._queue.clear()
+        self._changed()
+
+    def on_change(self, callback):
+        if not callable(callback):
+            raise TypeError("Only give callable arguments to 'on_change'")
+        self._callbacks.append(callback)
+
+    def replace(self, old, new):
+        if old not in self._queue:
+            raise ValueError("You can't replace unless one already exists, use add instead")
+        if self._queue[old] is old:
+            # re-insert with new name in old slot
+            self._replace_with_new_name(old, new)
+        else:
+            self._queue[old] = new
+        self._changed()
+
+    def remove(self, old):
+        if old not in self._queue:
+            raise ValueError("You can only remove something that has been added")
+        del self._queue[old]
+        self._changed()
```

Comment:
Just an idea: this could be done with a decorator.  I think the following should work if you like the pattern (I'm not sure it's actually better)

```python
def update_when_called(method):
    def inner(self, *args, **kwargs):
        return_value = method(self, *args, **kwargs)
        self._changed()
        return return_value
    return inner
```

Reply:
Slightly positive on it. Moot if we get rid of change listener. Punting, until that's cleared up.